nemo/lightning/pytorch/callbacks/memory_profiler.py

```diff
@@ -0,0 +1,58 @@
+import os
+from typing import List, Optional
+
+import torch
+from pytorch_lightning.callbacks.callback import Callback
+from nemo.lightning import io
+
+from nemo.utils import logging
+from nemo.utils.get_rank import get_rank
+
+
+class MemoryProfileCallback(Callback, io.IOMixin):
+    """
+    This callback enables recording a timeline of memory allocations during training.
+    The generated .pickle profiles can be analyzed at https://pytorch.org/memory_viz
+
+    More info about the profiles can be found [here](https://pytorch.org/blog/understanding-gpu-memory-1/).
+
+    Args:
+        dir (Optional[str]): Directory to store the memory profile dump
+
+    Example:
+        >>> callback = MemoryProfileCallback(dir="/mem_profile")
+        >>> trainer = Trainer(callbacks=[callback])
+    """
+
+    def __init__(self, dir: str = "/mem_profile"):
```

Comment:
Can we add an option to enable [`warn_tensor_cycles`](https://pytorch.org/blog/understanding-gpu-memory-2/) in this profiler as well?

Reply:
The blog post recommends adding that to model init. Is it possible to do this via a Callback? Could we just make this a flag for GPTModel?

Reply:
Addressed in latest commit

Reply:
I think we can just do it in the `on_train_start` method.

Reply:
Also might help catch issues before on_train_start